### PR TITLE
Divide expert gradients by the expert-parallel size

### PR DESCRIPTION
**Pull request by wujingyue**
Addresses #5656.

## What

Megatron-FSDP v2's `fully_shard` does **not** compose correctly with expert parallelism today: expert gradients come out `ep_size` times too large. This adds a `grad_divisor` argument that fixes it, plus the unit test that found it.

> An earlier revision of this PR claimed no library change was needed. That was wrong — see the measurements below.

## Why expert gradients are over-counted

A rank's experts process tokens routed to them from every rank in the expert-parallel group, and the backward pass routes those tokens' gradients back. So a rank's expert gradient already sums over `ep_size` ranks' data before any collective runs.

mFSDP then reduce-scatters with `ReduceOp.AVG` over the mesh it was handed. For experts that mesh is expert-DP (size `dp_size`), so it divides by `dp_size` when `world_size` ranks contributed — leaving expert gradients `ep_size` times too large. Dense parameters are unaffected: no alltoall, one contribution per mesh rank, and the full DP mesh is already the right size.

MCore DDP applies exactly this correction as `expert_gradient_scaling_factor = expt_dp / dp` (`distributed_data_parallel.py:208`), and Megatron-FSDP v1 carries it end to end. v2 had no way to express it, because the reduce op is chosen internally.

## Test

`tests/unit_tests/distributed/mfsdp_v2/test_expert_parallel.py` checks that an `EP=4` MoE `HybridModel` sharded with mFSDP v2 — experts over the expert-DP sub-mesh, dense params over the full DP mesh — consuming its `1/dp` shard of a global batch reproduces a single full-batch `EP=1` reference.

The reference has no distributed logic at all: it runs the whole batch on every rank with size-1 process groups, so its gradients are the exact full-batch gradients and need no reduction. That makes it an independent oracle for the alltoall dispatch, the sharding, and the gradient reduction together.

On 8×H100 (`EP=4`, `EDP=2`, `TP=PP=CP=1`) the model matches the reference to `max_abs=1.192093e-07` — `2^-23`, one ULP of float32. With `grad_divisor=1`, i.e. the pre-fix path, the same comparison peaks at `max_abs=1.578331e-03`: a 13,240× separation, so the test is a real regression guard.

The explicit `rtol=1e-3` / `atol=0` are dropped in favour of `assert_close`'s float32 defaults (`rtol=1.3e-6`, `atol=1e-5`).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: tests/unit_tests/distributed/mfsdp_v2/test_expert_parallel.py [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+        tp_cp=one,
+        tp_dp_cp=dp,
+        ep=ep,
+        tp_ep=ep,
+        expt_dp=expert_dp,
+        dp=dp,
+        dp_cp=dp,
+        embd=None,
+        pos_embd=None,
+    )
```
Comment (wujingyue) [inline tests/unit_tests/distributed/mfsdp_v2/test_expert_parallel.py:92]: @Connor-XY FYI, I had to `=one` many times when constructing a ProcessGroupCollection. Seems avoidable. 

(I tagged you because apparently you are trying to get rid of parallel_state in favor of ProcessGroupCollection)

File: tests/unit_tests/distributed/mfsdp_v2/test_expert_parallel.py [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+            for local, global_ in enumerate(model_layer.mlp.local_expert_indices):
+                getattr(model_fc, f"weight{local}").data.copy_(
+                    getattr(reference_fc, f"weight{global_}").data
+                )
+
+    # Shard the model: experts over the expert-DP sub-mesh, dense params over the full DP mesh.
+    for decoder_layer in model.decoder.layers:
+        if isinstance(decoder_layer, MoETransformerLayer):
+            fully_shard(decoder_layer.mlp.experts, mesh=moe_mesh["dp"], placements=_FLAT_SHARD)
+    fully_shard(model, mesh=world_mesh, placements=_FLAT_SHARD)
```
Comment (Autumn1998) [inline tests/unit_tests/distributed/mfsdp_v2/test_expert_parallel.py]: In MCore's terminology, `world_mesh` here is the DP mesh, and the `moe_mesh["dp"]` used
for the experts is the EDP mesh. Is this naming going to be confusing?

File: tests/unit_tests/distributed/mfsdp_v2/test_expert_parallel.py [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+            reference_fc = getattr(reference_layer.mlp.experts, fc)
+            for local, global_ in enumerate(model_layer.mlp.local_expert_indices):
+                getattr(model_fc, f"weight{local}").data.copy_(
+                    getattr(reference_fc, f"weight{global_}").data
+                )
+
+    # Shard the model: experts over the expert-DP sub-mesh, dense params over the full DP mesh.
+    for decoder_layer in model.decoder.layers:
+        if isinstance(decoder_layer, MoETransformerLayer):
+            fully_shard(decoder_layer.mlp.experts, mesh=moe_mesh["dp"], placements=_FLAT_SHARD)
```
Comment (Autumn1998) [inline tests/unit_tests/distributed/mfsdp_v2/test_expert_parallel.py]: Side question: do we support using ZeRO-1 for the MoE part and ZeRO-3 for the dense part?
If so, is there a corresponding unit test?

Comment (wujingyue) [reply tests/unit_tests/distributed/mfsdp_v2/test_expert_parallel.py]: I haven't tested ZeRO-1 at all. 

The placements passed into fully_shard should be 
```
model_weight=[Replicate]
main_grad=[Partial]
main_weight=[Shard]
```

The implementation should probably reuse what's added by #5743 and #5972. I don't expect more than 10 lines of code change in: https://github.com/NVIDIA/Megatron-LM/blob/076f61fc7be6dd0d8a7a95f99cc66dd775ac5c42/megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/parameter_group.py#L287. But I could certainly be wrong. 

Would you like to give it a shot?

File: tests/unit_tests/distributed/mfsdp_v2/test_expert_parallel.py (hunk 2 above)
Comment (wujingyue) [reply tests/unit_tests/distributed/mfsdp_v2/test_expert_parallel.py]: Do you prefer to call them `dp_mesh` and `edp_mesh`? 

There's a difference between "mesh" and "process group". A mesh is a container of process groups -- recall the `DeviceMesh.from_group` API. So I'd call a ProcessGroup `dp_group` or `edp_group` etc but keep calling mesh by their functionality (e.g. dense_mesh or moe_mesh). 

What do you think?

Comment (wujingyue) [inline tests/unit_tests/distributed/mfsdp_v2/test_expert_parallel.py]: Question for @Autumn1998 and @shjwudp: is this the sharding you care about for QWen 3.5? Basically, MoE are sharded on DP=4 and EP=2, and the rest on DP=8.

File: tests/unit_tests/distributed/mfsdp_v2/test_expert_parallel.py [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+        optimizer.zero_grad()
+        loss = torch.nn.functional.mse_loss(
+            model(input_ids=ids, position_ids=pos, attention_mask=mask), target
+        )
+        loss.backward()
+        optimizer.step()
+        # The model sees a shard, so average the loss across ranks for the global loss.
+        loss = loss.detach()
+        if loss_reduce_group is not None:
+            dist.all_reduce(loss, op=dist.ReduceOp.AVG, group=loss_reduce_group)
```
Comment (wujingyue) [inline tests/unit_tests/distributed/mfsdp_v2/test_expert_parallel.py:133]: Verify this is the right order. Should we all_reduce before loss.backward()?

File: tests/unit_tests/distributed/mfsdp_v2/test_expert_parallel.py [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+) -> list[torch.Tensor]:
+    """Run 5 SGD steps; return the per-step losses (globally averaged if loss_reduce_group given)."""
+    optimizer = torch.optim.SGD(model.parameters(), lr=0.02, foreach=False)
+    losses = []
+    for _ in range(5):
+        optimizer.zero_grad()
+        loss = torch.nn.functional.mse_loss(
+            model(input_ids=ids, position_ids=pos, attention_mask=mask), target
+        )
+        loss.backward()
```
Comment (wujingyue) [inline tests/unit_tests/distributed/mfsdp_v2/test_expert_parallel.py:128]: Double check: gradients may have to be divided by ep_size.

**Review by shjwudp — APPROVED** [review]

File: tests/unit_tests/distributed/mfsdp_v2/test_expert_parallel.py (hunk 3 above)
Comment (Autumn1998) [reply tests/unit_tests/distributed/mfsdp_v2/test_expert_parallel.py]: Ok, I will try no-shard/zero-1/zero-2

File: tests/unit_tests/distributed/mfsdp_v2/test_expert_parallel.py (hunk 2 above)
Comment (Autumn1998) [reply tests/unit_tests/distributed/mfsdp_v2/test_expert_parallel.py]: I mean, the name "dp" in ep_group, expert_dp_group = moe_mesh.get_group("ep"), moe_mesh.get_group("dp"), maybe use "edp" is more clear?

Comment (Autumn1998) [reply tests/unit_tests/distributed/mfsdp_v2/test_expert_parallel.py]: In the Qwen models, we are more focused on large models like the 397B class, which are very hard to fit within 8 ranks. Overall, we prefer using larger EP settings, such as EP4-EDP2 or EP8-EDP1. The latter in particular may introduce some corner cases.

Comment (wujingyue) [reply tests/unit_tests/distributed/mfsdp_v2/test_expert_parallel.py]: Certainly, will do -- thanks!

File: tests/unit_tests/distributed/mfsdp_v2/test_expert_parallel.py (hunk 4 above)
Comment (wujingyue) [reply tests/unit_tests/distributed/mfsdp_v2/test_expert_parallel.py:133]: @shjwudp The order seems correct as is. TorchTitan's training loop runs the following three steps in order: 
1. https://github.com/pytorch/torchtitan/blob/1c40dd26a91b0489249011a53d1da69061b03d38/torchtitan/trainer.py#L800. compute the local loss
2. https://github.com/pytorch/torchtitan/blob/1c40dd26a91b0489249011a53d1da69061b03d38/torchtitan/trainer.py#L845. compute the cross-rank averaged loss from the local loss
3. https://github.com/pytorch/torchtitan/blob/1c40dd26a91b0489249011a53d1da69061b03d38/torchtitan/trainer.py#L874. print the cross-rank averaged loss

File: tests/unit_tests/distributed/mfsdp_v2/test_expert_parallel.py (hunk 5 above)
Comment (wujingyue) [reply tests/unit_tests/distributed/mfsdp_v2/test_expert_parallel.py:128]: Thanks -- this turns out to be significant. I'll update the PR with the fix soon...

File: tests/unit_tests/distributed/mfsdp_v2/test_expert_parallel.py (hunk 4 above)
Comment (shjwudp) [reply tests/unit_tests/distributed/mfsdp_v2/test_expert_parallel.py:133]: My understanding is that this steps eventually determines the final metric loss. Assuming the loss values involved in the computation are numerically correct, it should be sufficient to inspect only the first step.

TorchTitan may compute the loss locally on a per-token basis. If we want to follow the per-token loss formulation more precisely, the calculation should be:

```python
local_loss_sum = sum(
    (per_token_loss * loss_mask).sum()
    for microbatch in accumulation_window
)
loss_for_backward = local_loss_sum * dp_world_size / global_tokens
loss_for_backward.backward()
```

Here, `per_token_loss` is the output of `cross_entropy(reduction="none")`.

Comment (wujingyue) [reply tests/unit_tests/distributed/mfsdp_v2/test_expert_parallel.py:133]: Are we talking about the same thing? I was trying to figure out whether the loss should be averaged **across ranks** before backward. As a data point, TorchTitan's training loop says no.

File: tests/unit_tests/distributed/mfsdp_v2/test_expert_parallel.py (hunk 5 above)
Comment (wujingyue) [reply tests/unit_tests/distributed/mfsdp_v2/test_expert_parallel.py:128]: Done

**Review by Phlip79 — APPROVED** [review]

File: tests/unit_tests/distributed/mfsdp_v2/test_expert_parallel.py (hunk 4 above)
Comment (shjwudp) [reply tests/unit_tests/distributed/mfsdp_v2/test_expert_parallel.py:133]: I think no need additional all-reduce of the loss is necessary for training correctness, since gradient averaging already provides the required effects.

If we assume that each DP rank receives an equivalent input, with no imbalance in the number of tokens, TorchTitan's approach is valid.

However, if the number of input tokens differs across DP ranks, we need to compute the loss on a per-token basis.

**Review by jiemingz — APPROVED** [review]

File: tests/unit_tests/distributed/mfsdp_v2/test_expert_parallel.py (hunk 4 above)
Comment (wujingyue) [reply tests/unit_tests/distributed/mfsdp_v2/test_expert_parallel.py:133]: > However, if the number of input tokens differs across DP ranks, we need to compute the loss on a per-token basis.

The local loss in this test is already locally normalized: mse_loss defaults to reduction="mean", averaging over the local batch × sequence × vocab elements. Every DP rank has the same shape and there is no loss mask, so averaging gradients across ranks is equivalent to computing the mean over the full global batch.

However, I agree that if the user computes the local loss as the sum we will need explicit global-token normalization. 

Am I understanding this correctly?

**Comment by chtruong814** [comment]
@wujingyue Some unit tests were failing in the merge queue now.  Would you please take a look?
https://github.com/NVIDIA/Megatron-LM/actions/runs/30503354295/job/90766469959

**Comment by wujingyue** [comment]
> **Edited twice.** This replaces my original analysis, which was wrong — it attributed the failures to NCCL reduction-order round-off and suggested loosening the test's tolerance. **Don't do that**; the assertion is catching a real bug. A later revision also wrongly claimed only `num_microbatches=3` was affected; corrected below. Earlier text is in this comment's edit history.

**Update on the mfsdp_v2 failures**

Comparing the sharded model's gradients against the baseline's at the end of step 0, before any optimizer step (every rank sees identical data, so a correct reduction must reproduce the baseline gradient exactly), with two spare 1-element all-reduces added to perturb timing:

| condition | gradients |
|---|---|
| unperturbed | exact — 48/48 comparisons zero, test passes 0/4 |
| + two 1-element all-reduces | **12%–125% wrong**, test fails 8/8 |

Worst observed: `layers.0.weight` off by 124.9%, `bias` by 67.8%. Magnitudes and the set of affected parameters vary run to run. The loss assertion only saw a ~1e-3 shadow of this because `lr=0.05` damps a one-step gradient error.

Both `num_microbatches` settings fail, `3` far more often than `1`. Counting failures of `test_hsdp_losses_match_baseline` across my runs: the PR arm gave 4 `[1-*]` and 3 `[3-*]`; a 32-run sweep gave 1 `[1-*]` and 35 `[3-*]`. The original CI failure here included `[1-True]` and `[1-False]`. So this is not specific to the deferred DP-outer accumulation path — with `num_microbatches=1` that path isn't exercised at all. More reductions per step just means more chances to hit it.

**It's the reduce-scatter stream.** Moving only `reduce_partial_gradients` to the current stream in `_reduce_gradient_groups` — leaving `partial_grad`'s allocation and both `wait_stream` edges untouched — makes it clean. So `main_grad` is written on `reduce_scatter_stream` and read somewhere without correct ordering.

I have **not** identified the missing ordering edge. Three hypotheses tested and refuted: `partial_grad` buffer lifetime (alloc and free are both on `reduce_scatter_stream`, so it's fine); autograd accumulating into `main_grad` (it doesn't — `copy_gradients_to_partial_buffer` reads the unsharded params); and the post-backward join resolving `current_stream()` on the autograd engine thread (adding a default-stream join neither fixes it nor is safe under CUDA graph capture).

Unperturbed mFSDP v2 gave bit-exact gradients in all my runs, so I have no evidence normal training is affected — but the perturbation needed to expose this is tiny.

PR 5956's source diff is a no-op for HSDP (at the default `grad_divisor=1` the new expression is identical to the old, and `test_fully_shard.py` never passes `grad_divisor`), this reproduces on clean main, and process groups are healthy throughout (598 known-answer collectives, 0 wrong). This looks like a pre-existing mFSDP v2 bug that 5956 surfaces by perturbing timing, not something 5956 introduces.

Reproducer: insert two `dist.all_reduce` calls on the mesh dim groups at the top of `test_hsdp_losses_match_baseline`, run the `mfsdp_v2` bucket on 8 GPUs. Fails 8/8. Ran on H200; CI is H100.
